File: src/field/Field.ts
import type {
  Cell,
  FieldConfig,
  GridPosition,
  Position,
  ScoringTarget,
  ZoneModifiers,
} from '../types/index.js';
import { ZoneType } from '../types/index.js';
import { Vector2D } from '../utils/Vector2D.js';

/**
 * Field class representing the game field as a grid
 */
export class Field {
  readonly config: FieldConfig;
  readonly cols: number;
  readonly rows: number;
  private readonly grid: Cell[][];

  constructor(config: FieldConfig) {
    this.config = config;
    this.cols = Math.ceil(config.width / config.cellSize);
    this.rows = Math.ceil(config.height / config.cellSize);
    this.grid = this.initializeGrid();
    this.applyZones();
  }

  /**
   * Initialize grid with default NORMAL cells
   */
  private initializeGrid(): Cell[][] {
    const grid: Cell[][] = [];
    for (let row = 0; row < this.rows; row++) {
      const rowCells: Cell[] = [];
      for (let col = 0; col < this.cols; col++) {
        rowCells.push({
          zone: ZoneType.NORMAL,
          modifiers: { speedMultiplier: 1.0 },
          baseCost: 1.0,
        });
      }
      grid.push(rowCells);
    }
    return grid;
  }

  /**
   * Apply zone definitions to the grid
   */
  private applyZones(): void {
    for (const zone of this.config.zones) {
      const minCol = Math.floor(zone.bounds.minX / this.config.cellSize);
      const maxCol = Math.ceil(zone.bounds.maxX / this.config.cellSize);
      const minRow = Math.floor(zone.bounds.minY / this.config.cellSize);
      const maxRow = Math.ceil(zone.bounds.maxY / this.config.cellSize);

      for (let row = minRow; row < maxRow && row < this.rows; row++) {
        for (let col = minCol; col < maxCol && col < this.cols; col++) {
          if (row >= 0 && col >= 0) {
            const cell = this.grid[row][col];
            cell.zone = zone.type;
            cell.modifiers = { ...cell.modifiers, ...zone.modifiers };
            cell.baseCost = this.calculateBaseCost(zone.type, zone.modifiers);
          }
        }
      }
    }
  }

  /**
   * Calculate base traversal cost for a zone type
   */
  private calculateBaseCost(type: ZoneType, modifiers?: ZoneModifiers): number {
    switch (type) {
      case ZoneType.NORMAL:
        return 1.0;
      case ZoneType.RAMP:
        return 1.0 / (modifiers?.speedMultiplier ?? 0.5);
      case ZoneType.TRENCH:
        return 1.1;
      case ZoneType.CLIMBING:
        return 1.5;
      case ZoneType.SCORING_ZONE:
        return 1.0;
      case ZoneType.OBSTACLE:
        return Infinity;
      case ZoneType.OUT_OF_BOUNDS:
        return Infinity;
      default:
        return 1.0;
    }
  }

  /**
   * Convert world position to grid position
   */
  positionToGrid(pos: Position): GridPosition {
    return {
      col: Math.floor(pos.x / this.config.cellSize),
      row: Math.floor(pos.y / this.config.cellSize),
    };
  }

  /**
   * Convert grid position to world position (center of cell)
   */
  gridToPosition(grid: GridPosition): Position {
    return {
      x: (grid.col + 0.5) * this.config.cellSize,
      y: (grid.row + 0.5) * this.config.cellSize,
    };
  }

  /**
   * Check if a grid position is within bounds
   */
  isInBounds(grid: GridPosition): boolean {
    return (
      grid.col >= 0 &&
      grid.col < this.cols &&
      grid.row >= 0 &&
      grid.row < this.rows
    );
  }

  /**
   * Check if a world position is within bounds
   */
  isPositionInBounds(pos: Position): boolean {
    return (
      pos.x >= 0 &&
      pos.x < this.config.width &&
      pos.y >= 0 &&
      pos.y < this.config.height
    );
  }

  /**
   * Get cell at grid position
   */
  getCell(grid: GridPosition): Cell | null {
    if (!this.isInBounds(grid)) return null;
    return this.grid[grid.row][grid.col];
  }

  /**
   * Get cell at world position
   */
  getCellAtPosition(pos: Position): Cell | null {
    return this.getCell(this.positionToGrid(pos));
  }

  /**
   * Get zone type at position
   */
  getZoneType(pos: Position): ZoneType {
    const cell = this.getCellAtPosition(pos);
    return cell?.zone ?? ZoneType.OUT_OF_BOUNDS;
  }

  /**
   * Get zone modifiers at position
   */
  getModifiers(pos: Position): ZoneModifiers {
    const cell = this.getCellAtPosition(pos);
    return cell?.modifiers ?? { speedMultiplier: 0 };
  }

  /**
   * Check if a position is in a no-score zone
   */
  isInNoScoreZone(pos: Position): boolean {
    const modifiers = this.getModifiers(pos);
    return modifiers.noScoring === true;
  }

  /**
   * Check if a position is in a ball-blocking zone (scoring areas block balls)
   */
  isInBallBlockingZone(pos: Position): boolean {
    const modifiers = this.getModifiers(pos);
    return modifiers.blocksBalls === true;
  }

  /**
   * Get ramp height at a position (for ball physics)
   * Returns 0 if not on a ramp, otherwise interpolated height based on position within ramp
   */
  getRampHeightAtPosition(pos: Position): number {
    const cell = this.getCellAtPosition(pos);
    if (!cell || cell.zone !== ZoneType.RAMP) return 0;

    const rampHeight = cell.modifiers.rampHeight ?? 0;
    if (rampHeight === 0) return 0;

    // Find the ramp zone this position is in
    for (const zone of this.config.zones) {
      if (zone.type !== 'RAMP') continue;
      if (
        pos.x >= zone.bounds.minX &&
        pos.x <= zone.bounds.maxX &&
        pos.y >= zone.bounds.minY &&
        pos.y <= zone.bounds.maxY
      ) {
        // Calculate position within ramp (0 = edge, 1 = center)
        const zoneCenterX = (zone.bounds.minX + zone.bounds.maxX) / 2;
        const zoneWidth = zone.bounds.maxX - zone.bounds.minX;
        const distFromCenter = Math.abs(pos.x - zoneCenterX);
        const normalizedDist = distFromCenter / (zoneWidth / 2);

        // Ramp is highest at center, tapers to 0 at edges
        // Use a smooth curve (parabolic) for the ramp profile
        const heightFactor = 1 - normalizedDist * normalizedDist;
        return rampHeight * Math.max(0, heightFactor);
      }
    }

    return 0;
  }

  /**
   * Check if a position is on a ramp
   */
  isOnRamp(pos: Position): boolean {
    const cell = this.getCellAtPosition(pos);
    return cell?.zone === ZoneType.RAMP;
  }

  /**
   * Check if there's a clear shot path between two positions (no obstacles/barriers)
   * Shots from a robot's own third of the field are allowed to pass into their scoring zone
   */
  hasClearShotPath(from: Position, to: Position, steps: number = 20): boolean {
    const fieldWidth = this.config.width;
    const leftThirdBoundary = fieldWidth / 3;      // 216" for 648" field
    const rightThirdBoundary = (fieldWidth * 2) / 3; // 432" for 648" field

    // Determine which third the shot originates from
    const isFromLeftThird = from.x < leftThirdBoundary;
    const isFromRightThird = from.x > rightThirdBoundary;

    for (let i = 1; i < steps; i++) {
      const t = i / steps;
      const checkPos: Position = {
        x: from.x + (to.x - from.x) * t,
        y: from.y + (to.y - from.y) * t,
      };
      const cell = this.getCellAtPosition(checkPos);
      if (!cell) return false;

      // Check for obstacles that block shots
      if (cell.zone === ZoneType.OBSTACLE) return false;

      // Cells with blocksBalls modifier block shots that pass through them,
      // UNLESS the shot originates from the proper third of the field
      // (Check blocksBalls modifier rather than zone type since zones can be overwritten)
      if (cell.modifiers.blocksBalls) {
        // Check if this blocking zone is on the left or right side
        const zoneIsOnLeftSide = checkPos.x < fieldWidth / 2;

        // Allow shots from left third through left blocking zone (red side)
        // Allow shots from right third through right blocking zone (blue side)
        const isAllowed = (isFromLeftThird && zoneIsOnLeftSide) ||
                          (isFromRightThird && !zoneIsOnLeftSide);

        if (!isAllowed) {
          return false;
        }
      }
    }
    return true;
  }

  /**
   * Check if a position is good for shooting (not on ramp, not in trench, not in no-score zone)
   */
  isGoodShootingPosition(pos: Position): boolean {
    const cell = this.getCellAtPosition(pos);
    if (!cell) return false;

    // Can't shoot from ramps
    if (cell.zone === ZoneType.RAMP) return false;

    // Can't shoot from trenches
    if (cell.zone === ZoneType.TRENCH) return false;

    // Can't shoot from no-score zone
    if (cell.modifiers.noScoring) return false;

    return true;
  }

  /**
   * Check if a position is on the correct side of the field for an alliance
   */
  isOnAllianceSide(pos: Position, alliance: 'red' | 'blue'): boolean {
    const centerX = this.config.width / 2;
    // Red is on the left (low X), Blue is on the right (high X)
    if (alliance === 'red') {
      return pos.x < centerX;
    } else {
      return pos.x > centerX;
    }
  }

  /**
   * Check if a robot can traverse a cell based on height restrictions
   */
  canRobotTraverse(grid: GridPosition, robotHeight: number): boolean {
    const cell = this.getCell(grid);
    if (!cell) return false;

    // Check for impassable zones
    if (cell.zone === ZoneType.OBSTACLE || cell.zone === ZoneType.OUT_OF_BOUNDS) {
      return false;
    }

    // Scoring zones with blocksBalls are also impassable to robots
    if (cell.zone === ZoneType.SCORING_ZONE && cell.modifiers.blocksBalls) {
      return false;
    }

    // Check height restrictions (e.g., trench)
    if (cell.modifiers.maxHeight !== undefined) {
      if (robotHeight > cell.modifiers.maxHeight) {
        return false;
      }
    }

    // Check for protected zones
    if (cell.modifiers.protected) {
      return false;
    }

    return true;
  }

  /**
   * Get traversal cost for a robot at a grid position
   */
  getTraversalCost(
    grid: GridPosition,
    robotHeight: number,
    _robotAlliance?: 'red' | 'blue'
  ): number {
    const cell = this.getCell(grid);
    if (!cell) return Infinity;

    if (!this.canRobotTraverse(grid, robotHeight)) {
      return Infinity;
    }

    return cell.baseCost;
  }

  /**
   * Get neighbors of a grid position (8-directional)
   */
  getNeighbors(grid: GridPosition): GridPosition[] {
    const neighbors: GridPosition[] = [];
    const directions = [
      { col: -1, row: -1 },
      { col: 0, row: -1 },
      { col: 1, row: -1 },
      { col: -1, row: 0 },
      { col: 1, row: 0 },
      { col: -1, row: 1 },
      { col: 0, row: 1 },
      { col: 1, row: 1 },
    ];

    for (const dir of directions) {
      const neighbor = {
        col: grid.col + dir.col,
        row: grid.row + dir.row,
      };
      if (this.isInBounds(neighbor)) {
        neighbors.push(neighbor);
      }
    }

    return neighbors;
  }

  /**
   * Get scoring targets for an alliance
   */
  getScoringTargets(alliance: 'red' | 'blue'): ScoringTarget[] {
    return this.config.scoringTargets.filter((t) => t.alliance === alliance);
  }

  /**
   * Find the nearest scoring target to a position
   */
  getNearestScoringTarget(
    pos: Position,
    alliance: 'red' | 'blue'
  ): ScoringTarget | null {
    const targets = this.getScoringTargets(alliance);
    if (targets.length === 0) return null;

    const position = Vector2D.fromPosition(pos);
    let nearest: ScoringTarget | null = null;
    let nearestDist = Infinity;

    for (const target of targets) {
      const dist = position.distanceTo(target.position);
      if (dist < nearestDist) {
        nearestDist = dist;
        nearest = target;
      }
    }

    return nearest;
  }

  /**
   * Check if a position is in range of a scoring target
   */
  isInScoringRange(
    pos: Position,
    targetId: string,
    shootingRange: number
  ): boolean {
    const target = this.config.scoringTargets.find((t) => t.id === targetId);
    if (!target) return false;

    const distance = Vector2D.fromPosition(pos).distanceTo(target.position);
    return distance <= shootingRange;
  }

  /**
   * Get starting positions for an alliance
   */
  getStartingPositions(alliance: 'red' | 'blue'): Position[] {
    return this.config.startingPositions[alliance];
  }

  /**
   * Get ball spawn points
   */
  getBallSpawnPoints(): Position[] {
    return this.config.ballSpawnPoints.map((sp) => sp.position);
  }

  /**
   * Check if a position is within or near a climbing zone for the specified alliance
   */
  isNearClimbingZone(pos: Position, alliance: 'red' | 'blue', proximityThreshold: number = 36): boolean {
    // Find climbing zones for this alliance
    for (const zone of this.config.zones) {
      if (zone.type !== ZoneType.CLIMBING) continue;

      // Check if this is the alliance's climbing zone
      // Red climbing zone is on the left (low X), Blue is on the right (high X)
      const isRedZone = zone.bounds.minX < this.config.width / 2;
      const isAllianceZone = (alliance === 'red' && isRedZone) || (alliance === 'blue' && !isRedZone);

      if (!isAllianceZone) continue;

      // Check if position is within the zone or within proximity threshold
      const inZoneX = pos.x >= zone.bounds.minX && pos.x <= zone.bounds.maxX;
      const inZoneY = pos.y >= zone.bounds.minY && pos.y <= zone.bounds.maxY;

      if (inZoneX && inZoneY) return true;

      // Check proximity to zone bounds
      const distToZone = this.distanceToZone(pos, zone.bounds);
      if (distToZone <= proximityThreshold) return true;
    }

    return false;
  }

  /**
   * Calculate minimum distance from a position to a zone's bounds
   */
  private distanceToZone(pos: Position, bounds: { minX: number; maxX: number; minY: number; maxY: number }): number {
    const clampedX = Math.max(bounds.minX, Math.min(bounds.maxX, pos.x));
    const clampedY = Math.max(bounds.minY, Math.min(bounds.maxY, pos.y));
    const dx = pos.x - clampedX;
    const dy = pos.y - clampedY;
    return Math.sqrt(dx * dx + dy * dy);
  }
}

File: src/frc_sim/field/field.py
"""Field representation with zone management."""

from __future__ import annotations
import math
from typing import Optional

from ..types.enums import ZoneType
from ..types.schemas import (
    Position,
    FieldConfig,
    ZoneDefinition,
    ScoringTarget,
    BallSpawnPoint,
)


class Field:
    """Field representation with zone grid and utilities."""

    def __init__(self, config: FieldConfig):
        self.config = config
        self._zone_cache: dict[tuple[int, int], ZoneDefinition | None] = {}

    @property
    def width(self) -> float:
        """Get field width."""
        return self.config.width

    @property
    def height(self) -> float:
        """Get field height."""
        return self.config.height

    def get_zone_at(self, pos: Position) -> Optional[ZoneDefinition]:
        """Get the zone at a position."""
        # Use integer grid position for caching
        grid_key = (int(pos.x), int(pos.y))
        if grid_key in self._zone_cache:
            return self._zone_cache[grid_key]

        for zone in self.config.zones:
            bounds = zone.bounds
            if (
                bounds.minX <= pos.x <= bounds.maxX
                and bounds.minY <= pos.y <= bounds.maxY
            ):
                self._zone_cache[grid_key] = zone
                return zone

        self._zone_cache[grid_key] = None
        return None

    def get_zone_type_at(self, pos: Position) -> ZoneType:
        """Get the zone type at a position."""
        zone = self.get_zone_at(pos)
        if zone:
            return ZoneType(zone.type)
        return ZoneType.NORMAL

    def get_speed_multiplier_at(self, pos: Position) -> float:
        """Get the speed multiplier at a position."""
        zone = self.get_zone_at(pos)
        if zone and zone.modifiers and zone.modifiers.speedMultiplier is not None:
            return zone.modifiers.speedMultiplier
        return 1.0

    def get_ramp_height_at(self, pos: Position) -> float:
        """Get the ramp height at a position.

        Uses parabolic profile to match TypeScript implementation.
        Ramp is highest at center, tapers to 0 at edges.
        """
        zone = self.get_zone_at(pos)
        if zone and zone.modifiers and zone.modifiers.rampHeight is not None:
            ramp_height = zone.modifiers.rampHeight
            if ramp_height == 0:
                return 0.0

            # Calculate position within ramp (0 = edge, 1 = center)
            bounds = zone.bounds
            zone_center_x = (bounds.minX + bounds.maxX) / 2
            zone_width = bounds.maxX - bounds.minX

            if zone_width == 0:
                return ramp_height

            dist_from_center = abs(pos.x - zone_center_x)
            normalized_dist = dist_from_center / (zone_width / 2)

            # Ramp is highest at center, tapers to 0 at edges
            # Use a smooth curve (parabolic) for the ramp profile
            height_factor = 1 - normalized_dist * normalized_dist
            return ramp_height * max(0.0, height_factor)

        return 0.0

    def is_in_bounds(self, pos: Position) -> bool:
        """Check if position is in bounds."""
        return 0 <= pos.x <= self.config.width and 0 <= pos.y <= self.config.height

    def is_passable(self, pos: Position, robot_height: float = 0.0) -> bool:
        """Check if a position is passable for a robot."""
        if not self.is_in_bounds(pos):
            return False

        zone = self.get_zone_at(pos)
        if zone:
            # Check for impassable zones
            if zone.type == ZoneType.OBSTACLE.value:
                return False
            if zone.type == ZoneType.OUT_OF_BOUNDS.value:
                return False

            # Scoring zones with blocksBalls are also impassable to robots
            if (
                zone.type == ZoneType.SCORING_ZONE.value
                and zone.modifiers
                and zone.modifiers.blocksBalls
            ):
                return False

            if zone.modifiers:
                # Check for protected zones
                if zone.modifiers.protected:
                    return False
                # Check height restrictions (e.g., trench)
                if (
                    zone.modifiers.maxHeight is not None
                    and robot_height > zone.modifiers.maxHeight
                ):
                    return False
        return True

    def get_cost_at(self, pos: Position) -> float:
        """Get the pathfinding cost at a position.

        Matches TypeScript calculateBaseCost implementation.
        """
        zone = self.get_zone_at(pos)
        if zone:
            zone_type = zone.type
            modifiers = zone.modifiers

            if zone_type == ZoneType.NORMAL.value:
                return 1.0
            if zone_type == ZoneType.RAMP.value:
                # Use inverse of speedMultiplier (default 0.5 -> cost 2.0)
                speed_mult = modifiers.speedMultiplier if modifiers else 0.5
                return 1.0 / (speed_mult if speed_mult else 0.5)
            if zone_type == ZoneType.TRENCH.value:
                return 1.1
            if zone_type == ZoneType.CLIMBING.value:
                return 1.5
            if zone_type == ZoneType.SCORING_ZONE.value:
                return 1.0
            if zone_type == ZoneType.OBSTACLE.value:
                return float("inf")
            if zone_type == ZoneType.OUT_OF_BOUNDS.value:
                return float("inf")

        return 1.0

    def is_near_climbing_zone(
        self, pos: Position, alliance: str, proximity_threshold: float = 36.0
    ) -> bool:
        """Check if position is within or near a climbing zone for the alliance.

        Uses position-based alliance detection: red climbing zone is on the left
        (low X), blue is on the right (high X).
        """
        for zone in self.config.zones:
            if zone.type != ZoneType.CLIMBING.value:
                continue

            # Check if this is the alliance's climbing zone based on position
            # Red climbing zone is on the left (low X), Blue is on the right (high X)
            is_red_zone = zone.bounds.minX < self.config.width / 2
            is_alliance_zone = (alliance == "red" and is_red_zone) or (
                alliance == "blue" and not is_red_zone
            )

            if not is_alliance_zone:
                continue

            # Check if position is within the zone
            bounds = zone.bounds
            in_zone_x = bounds.minX <= pos.x <= bounds.maxX
            in_zone_y = bounds.minY <= pos.y <= bounds.maxY

            if in_zone_x and in_zone_y:
                return True

            # Check proximity to zone bounds using clamped distance
            dist_to_zone = self._distance_to_zone(pos, bounds)
            if dist_to_zone <= proximity_threshold:
                return True

        return False

    def _distance_to_zone(
        self,
        pos: Position,
        bounds: "ZoneDefinition.bounds",
    ) -> float:
        """Calculate minimum distance from a position to a zone's bounds."""
        clamped_x = max(bounds.minX, min(bounds.maxX, pos.x))
        clamped_y = max(bounds.minY, min(bounds.maxY, pos.y))
        dx = pos.x - clamped_x
        dy = pos.y - clamped_y
        return math.sqrt(dx * dx + dy * dy)

    def is_in_no_score_zone(self, pos: Position) -> bool:
        """Check if position is in a no-scoring zone."""
        zone = self.get_zone_at(pos)
        if zone and zone.modifiers and zone.modifiers.noScoring:
            return True
        return False

    def is_in_ball_blocking_zone(self, pos: Position) -> bool:
        """Check if position is in a ball-blocking zone."""
        zone = self.get_zone_at(pos)
        if zone and zone.modifiers and zone.modifiers.blocksBalls:
            return True
        return False

    def get_starting_positions(self, alliance: str) -> list[Position]:
        """Get starting positions for an alliance."""
        if alliance == "red":
            return self.config.startingPositions.red
        return self.config.startingPositions.blue

    def get_scoring_targets(self, alliance: str) -> list[ScoringTarget]:
        """Get scoring targets for an alliance."""
        return [t for t in self.config.scoringTargets if t.alliance == alliance]

    def get_scoring_target(self, target_id: str) -> Optional[ScoringTarget]:
        """Get a scoring target by ID."""
        for target in self.config.scoringTargets:
            if target.id == target_id:
                return target
        return None

    def get_ball_spawn_points(self) -> list[BallSpawnPoint]:
        """Get all ball spawn points."""
        return self.config.ballSpawnPoints

    def get_spawn_point(self, spawn_id: str) -> Optional[BallSpawnPoint]:
        """Get a spawn point by ID."""
        for sp in self.config.ballSpawnPoints:
            if sp.id == spawn_id:
                return sp
        return None

    def is_on_alliance_side(self, pos: Position, alliance: str) -> bool:
        """Check if position is on the alliance's side of the field."""
        mid_x = self.config.width / 2
        if alliance == "red":
            return pos.x < mid_x
        return pos.x >= mid_x

    def is_good_shooting_position(self, pos: Position) -> bool:
        """Check if position is good for shooting.

        Cannot shoot from ramps, trenches, or no-score zones.
        """
        zone = self.get_zone_at(pos)

        # Can't shoot from ramps
        if zone and zone.type == ZoneType.RAMP.value:
            return False

        # Can't shoot from trenches
        if zone and zone.type == ZoneType.TRENCH.value:
            return False

        # Can't shoot from no-score zone
        if self.is_in_no_score_zone(pos):
            return False

        return True

    def has_clear_shot_path(self, from_pos: Position, to_pos: Position) -> bool:
        """Check if there's a clear shot path between two positions.

        Currently returns True as we don't have obstacle detection for shots.
        This could be extended to check for obstacles along the path.
        """
        # For now, assume all shots have a clear path
        # A more sophisticated implementation could check for obstacles
        return True

    def clamp_to_bounds(self, pos: Position, margin: float = 0.0) -> Position:
        """Clamp a position to be within field bounds."""
        return Position(
            x=max(margin, min(self.config.width - margin, pos.x)),
            y=max(margin, min(self.config.height - margin, pos.y)),
        )
